File: DAYS/Day05-MaxSubArray/MaxSubArray.js
// Array problem: Maximum Subarray
// Given an integer array nums, find the contiguous subarray 
// (containing at least one number) which has the largest sum and return its sum.


function maxSubArray(nums){
    let maxSum=nums[0]; // Initialize maximum sum to the first element of the array
    let curSum=0; // Initialize current sum to 0
 
    for(let num of nums){
        curSum += num // Add the current element to the current sum
        maxSum=Math.max(maxSum,curSum); // Update the maximum sum if the current sum is greater
        if (curSum<0){ // If the current sum is negative, reset it to 0
            curSum = 0;
        }
    }
    return maxSum; // Return the maximum subarray sum
}

// Test cases
console.log(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))
// output: 6
console.log(maxSubArray([1]))
// output: 1
console.log(maxSubArray([5,4,-1,7,8]))
// output: 23
console.log(maxSubArray([-1]))
// output: -1

// Best Approach: Kadane's Algorithm
// complexity Analysis : Time complexity: O(n) ,Space complexity: O(1)

File: DAYS/Day05-MaxSubArray/MaxSubArray.py
# Array Problem: Maximum Subarray
# Given an integer array nums, find the contiguous subarray (containing at least one number) 
# which has the largest sum and return its sum.


def maxSubArray(nums):
    # Initialize maxSum and curSum with the first element of nums
    maxSum =nums[0]
    curSum =0
    for num in nums:
        # Add the current element to curSum
        curSum += num
        # Update maxSum if curSum is greater
        maxSum = max(maxSum, curSum)
        # Reset curSum if it is negative
        if curSum < 0:
            curSum = 0
    return maxSum

# Test cases
print(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))
# Output: 6
print(maxSubArray([1]))
# Output: 1
print(maxSubArray([5,4,-1,7,8]))
# Output: 23

# Best Approach: Kadane's Algorithm
# complexity Analysis: time complexity: O(n), space complexity: O(1)